plugin.video.s04tv/default.py

```diff
@@ -0,0 +1,369 @@
+        
+        if(mode == 1):
+            addDir(title, url, mode, imageUrl)
+        else:
+            addLink(title, url, mode, imageUrl, '', extraInfo)
+
+
+def getVideoUrl(url, doc):
+    xbmc.log('getVideoUrl: url=' +url)
+
+    is_logged_in, has_schalke_tv = login()
+    if not is_logged_in:
+        return
+
+    isPayedContent = xbmc.getInfoLabel( "ListItem.Property(isPayedContent)" ) == 'True'
+    #payedContent requires payed Schalke TV subscription
+    if(isPayedContent):
+        if not has_schalke_tv:
+            xbmc.log("This video requires payed Schalke TV subscription")
+            xbmcgui.Dialog().ok(PLUGINNAME, language(30104), language(30105))
+            return
+
+    
+    #HACK: Free content may be hosted on youtube
+    if(url.startswith("https://youtu.be/")):
+        videoId = url.replace("https://youtu.be/", "")
+        url='plugin://plugin.video.youtube/?action=play_video&videoid=' +videoId
+        listitem = xbmcgui.ListItem(path=url)
+        return xbmcplugin.setResolvedUrl(thisPlugin, True, listitem)
+    
+    soup = BeautifulSoup(''.join(doc))
+    
+    section = soup.find('section')
+    if(not section):
+        xbmc.log(missingelementtext%'section')
+        return
+    
+    videoContainer = soup.find('div', attrs={'class': 'video-wrapper'})
+    if(not videoContainer):
+        xbmc.log(missingelementtext%'video-wrapper')
+        return
+    
+    dataoptions = videoContainer['data-options']
+    if(not dataoptions):
+        xbmc.log(missingelementtext%'data-options')
+        return
+        
+    jsonResult = json.loads(dataoptions)
+    jsonResultPlayerOptions = jsonResult['playerOptions']
+    dataId = jsonResultPlayerOptions['data-id']
+    
+    playoutUrl = 'https://playout.3qsdn.com/' +dataId +'?js=true&skin=s04&data-id=' +dataId +'&container=sdnPlayer_player&preview=false&width=100%25&height=100%25'
+    
+    response=getUrl(playoutUrl)
+    
+    match_playlist=re.compile('playlist: \((.+?)\)', re.DOTALL).findall(response)
+    playlist = match_playlist[0]
+    
+    quote_keys_regex = r'([\{\s,])(\w+)(:)'
+    playlist = re.sub(quote_keys_regex, r'\1"\2"\3', playlist)
+
+    playlist = playlist.replace("'", '"')
+    playlist = playlist.replace("\\x2F", '')
+    
+    videourl = ''
+    
+    jsonPlaylist = json.loads(playlist)
+    for key in jsonPlaylist:
+        entry = jsonPlaylist[key]
+        quality = entry['quality']
+        videotype = entry['type']
+        if(quality == videoquality):
+            videourl = entry['src']
+            
+    listitem = xbmcgui.ListItem(path=videourl)
+    return xbmcplugin.setResolvedUrl(thisPlugin, True, listitem)
+    
+
+def addDir(name, url, mode, iconimage):
+    parameters = {'url' : url.encode('utf-8'), 'mode' : str(mode), 'name' : name.encode('utf-8')}
+    u = sys.argv[0] +'?' +urllib.urlencode(parameters)
+    ok = True
+    listitem = xbmcgui.ListItem(name, iconImage="DefaultFolder.png", thumbnailImage=iconimage)
```

Comment:
iconImage and thumbnailImage are both deprecated, please use listitem.setArt{"icon":"xxx", "thumb":"..."}